### PR TITLE
Álvaro Day 1

### PR DESCRIPTION
**Pull request by alvarocasadoc**
I've solved this first problem with Elixir. I have some concerns about my solution's complexity, but I think it's not that bad for a first look into functional programming.

## Structure
- My solution is located in the `alvarocasadoc/day1/day1.exs` file. 
- A simple demo can be found at `alvarocasadoc/day1/day1_demo.exs`.
- There are some tests in `alvarocasadoc/day1/day1_test.exs`.

## Part 1. Sum all elements from a list
I've defined the following function:

```elixir
  def sum_frequencies(input) do
    items = Enum.map(input, fn x -> Integer.parse(x) |> elem(0) end)
    Enum.sum(items)
  end
```
## Part 2. Find first recurring sum result in a cycling list
I've managed to solve it using Streams and the `Enum.reduce_while` function, which I did not know nothing about before.

```elixir
  def find_first_cycle(input) do
    items = Enum.map(input, fn x -> Integer.parse(x) |> elem(0) end)
    stream = Stream.scan(Stream.cycle(items), &(&1 + &2))

    Enum.reduce_while(stream, [], fn current, prev ->
      if current in prev, do: {:halt, current}, else: {:cont, [current | prev]}
    end)
  end
```

## Tests
image

### CODE REVIEW

File: alvarocasadoc/day1/day1.exs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# AdventOfCode2018
+# @alvarocasadoc
+# Day 1
+
+defmodule Day1 do
+
+  # Part 1
+  def sum_frequencies(input) do
+    items = Enum.map(input, fn x -> Integer.parse(x) |> elem(0) end)
+    Enum.sum(items)
```
Comment (Cotel) [inline alvarocasadoc/day1/day1.exs:10]: Okay, so if I understand correctly the functions for collections are inside the module `Enum`?

Then, as Integer.parse returns a Tuple, the `|>` thing is `then`?

If that's so you could do which is still readable and keeps the function more `mathematical` (but that's personal preference)

```elixir
def sum_frequencies(input) do
  Enum.map(input, fn x -> Integer.parse(x) |> elem(0) end) |> Enum.sum()
end
```

Good job 👍 I really like this

File: alvarocasadoc/day1/day1.exs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  end
+
+  # Part 2
+  def find_first_cycle(input) do
+    items = Enum.map(input, fn x -> Integer.parse(x) |> elem(0) end)
```
Comment (Cotel) [inline alvarocasadoc/day1/day1.exs:15]: Do you see a pattern here? 😏 You can extract a function for parsing

File: alvarocasadoc/day1/day1.exs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    stream = Stream.scan(Stream.cycle(items), &(&1 + &2))
```
Comment (Cotel) [inline alvarocasadoc/day1/day1.exs:16]: The `&` thingy is a little bit confusing. I suppose it is some kind of compact lambda and you are just telling to sum elements. Can you substitute that for a function reference like `Stream.scan(..., sum)`?

File: alvarocasadoc/day1/day1.exs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+    Enum.reduce_while(stream, [], fn current, prev ->
+      if current in prev, do: {:halt, current}, else: {:cont, [current | prev]}
+    end)
```
Comment (Cotel) [inline alvarocasadoc/day1/day1.exs:20]: Hmmm this kind of syntax is why I don't like Ruby or Elixir that much (though they are pretty cool).

It is pretty clear what this is doing, but still, what are `:halt` and `:cont` exactly? I suppose this is some kind of trick they are using internally but I would prefer some other way of expressing I want to stop or continue. Maybe a lambda or a keywork like `return` dunno.

Anyway, this function is pretty clever. I would say this is even more declarative than stopping recursion in Haskell 😇

**Review by Cotel — APPROVED** [review]
Amazing! 💯 I hope you can participate in the challenge for a few more days so I can see more Elixir 👍

File: alvarocasadoc/day1/day1.exs (hunk 1 above)
Comment (alvarocasadoc) [reply alvarocasadoc/day1/day1.exs:10]: Yes, you're right! `|>` works as a pipe operator.

**Pull request merged**